File: frontend/src/api/biliPassport.ts
/**
 * `/bili-passport/*` -- B 站扫码登录 + 身份码自动获取.
 */

import { ENV } from "@/config/env";
import { api, toApiError } from "./http";

const url = (path: string) => ENV.BASE_PATH + "/bili-passport" + path;

export interface QrcodeResp {
    code: number;
    url?: string;
    qrcode_key?: string;
    message?: string;
}

export interface PollResp {
    /** 0=登录成功 / 86038=失效 / 86090=已扫待确认 / 86101=未扫 / -1=网络错. */
    code: number;
    message?: string;
    /** 登录成功时返回的 cookie 串; 前端应当持久化到 sharedConfig. */
    cookie?: string;
}

export interface WhoamiResp {
    logged: boolean;
    mid?: number;
    uname?: string;
    avatar?: string;
}

export const biliPassportApi = {
    async generateQrcode(): Promise<QrcodeResp> {
        try {
            const { data } = await api.get<QrcodeResp>(url("/qrcode"), { timeout: 8000 });
            return data;
        } catch (err) {
            return { code: -1, message: toApiError(err).message };
        }
    },

    async poll(qrcodeKey: string): Promise<PollResp> {
        try {
            const { data } = await api.get<PollResp>(url("/poll"), {
                params: { qrcode_key: qrcodeKey },
                timeout: 6000,
            });
            return data;
        } catch (err) {
            return { code: -1, message: toApiError(err).message };
        }
    },

    /** 应用启动 hydrate 完 cookie 后调一次, 把 cookie 写回后端内存. */
    async setCookie(cookie: string): Promise<boolean> {
        try {
            await api.post(url("/setCookie"), { cookie }, { timeout: 4000 });
            return true;
        } catch (err) {
            console.warn("[bili-passport] setCookie 失败：", err);
            return false;
        }
    },

    async whoami(): Promise<WhoamiResp> {
        try {
            const { data } = await api.get<WhoamiResp>(url("/whoami"), { timeout: 6000 });
            return data;
        } catch {
            return { logged: false };
        }
    },

    async logout(): Promise<void> {
        try {
            await api.post(url("/logout"), {}, { timeout: 4000 });
        } catch {
            /* ignore */
        }
    },
};

File: desktop-tauri/src-tauri/crates/server/src/routes/bili_passport.rs
//! B 站扫码登录路由

use axum::{
    extract::{Query, State},
    response::IntoResponse,
    routing::{get, post},
    Json, Router,
};
use ordersong_bili_passport::{qrcode, whoami};
use serde::Deserialize;
use serde_json::{json, Value};

use crate::error::{ApiError, ApiResult};
use crate::state::AppStateRef;

pub fn router() -> Router<AppStateRef> {
    Router::new()
        .route("/qrcode", get(qrcode_generate))
        .route("/poll", get(qrcode_poll))
        .route("/setCookie", post(set_cookie_manual))
        .route("/whoami", get(whoami_handler))
        .route("/logout", post(logout))
}

async fn qrcode_generate(State(state): State<AppStateRef>) -> impl IntoResponse {
    match qrcode::generate(&state.http).await {
        Ok(qr) => Json(json!({ "code": 0, "url": qr.url, "qrcode_key": qr.qrcode_key })),
        Err(msg) => Json(json!({ "code": -1, "message": msg })),
    }
}

#[derive(Deserialize)]
struct PollQuery {
    qrcode_key: String,
}

async fn qrcode_poll(State(state): State<AppStateRef>, Query(q): Query<PollQuery>) -> impl IntoResponse {
    match qrcode::poll(&state.http, &q.qrcode_key).await {
        Ok(r) => {
            if let Some(cookie) = r.cookie {
                if let Ok(mut w) = state.bili_cookie.write() {
                    *w = cookie.clone();
                }
                Json(json!({ "code": 0, "message": "登录成功", "cookie": cookie }))
            } else {
                Json(json!({ "code": r.code, "message": r.message }))
            }
        }
        Err(msg) => Json(json!({ "code": -1, "message": msg })),
    }
}

#[derive(Deserialize)]
struct SetCookieBody {
    cookie: Option<String>,
}

async fn set_cookie_manual(
    State(state): State<AppStateRef>,
    Json(b): Json<SetCookieBody>,
) -> ApiResult<Json<Value>> {
    let c = b.cookie.unwrap_or_default();
    let mut w = state
        .bili_cookie
        .write()
        .map_err(|_| ApiError::internal("B 站 cookie 写锁中毒"))?;
    *w = c;
    Ok(Json(json!({ "code": 0 })))
}

async fn logout(State(state): State<AppStateRef>) -> impl IntoResponse {
    if let Ok(mut w) = state.bili_cookie.write() {
        w.clear();
    }
    Json(json!({ "code": 0 }))
}

async fn whoami_handler(State(state): State<AppStateRef>) -> impl IntoResponse {
    let cookie = state.bili_cookie_snapshot();
    let info = whoami::whoami(&state.http, &cookie).await;
    Json(json!({
        "logged": info.logged,
        "mid": info.mid,
        "uname": info.uname,
        "avatar": info.avatar,
    }))
}
